### ISSUE TITLE
Misleading description of how Julia uses types

### ISSUE THREAD

**Issue by jiahao** [tpetricek/TomaspNet.Website#16]
The post  "Against Types" contains the statement

> Julia uses types almost exclusively for performance.

However, this statement is rather misleading, because performance gains don't come magically just from specifying types.

From a programming language semantics point of view, Julia uses types purely for method dispatch of generic functions, and nothing else. For example, `a * b` currently could mean string concatenation, matrix product, or complex scalar product, and the types of the arguments are used to decide which meaning of `*` is intended. At this level, types have nothing to do with performance.

Nevertheless, it is true that types can indirectly lead to performance gains, when at least one other condition holds:

1. When users (or library writers) write specialized methods of a generic function that implement optimized algorithms, and users write code that dispatches on those specialized methods.

2. When the JIT compiler deduces enough type information to eliminate the overhead of dynamic method dispatch. This happens when static analyses like data flow type inference, function inlining and tuple elision reveal enough type information to determine the method being dispatched uniquely, or at least reduce the dispatch choices to sufficiently few cases that the compiler may choose to perform aggressive method specialization for all the remaining cases.

3. When the JIT compiler deduces enough type information to infer concrete types for identifiers, enabling further compiler optimizations such as constant folding and unboxing immutable types.

The first case requires users to use the generic function system to define algorithmic specializations. The latter two are implementation specific, and the performance characteristics depend on how smart the compiler is.

In summary, using types in Julia is neither necessary nor sufficient to guarantee performance, but it does help when at least one of these three cases is triggered.

For more information, you may be interested in @JeffBezanson's PhD thesis: https://github.com/JeffBezanson/phdthesis

**Comment by StefanKarpinski** [comment]
Btw, thank you for writing this essay. I thoroughly enjoyed it, generally agree with your position, and feel strongly that we need more people who are willing to look at the concept of types beyond the confines of the PFPL-approved definition.

Regarding Julia specifically, it does use types in a number of ways other than for performance, including:

- detecting errors dynamically (but not statically, by default)
- abstraction and support for structuring large systems
- describing the representation of data
- expressiveness (multiple dispatch)
- documentation

Notably not for whole language safety, of course. Or as [Leah Hanson](https://github.com/astrieanna) once eloquently put it, "Julia uses the type system in all the ways that don't end with the programmer arguing with the compiler."

**Comment by JeffBezanson** [comment]
Seconded! Great essay. I cited Stephen Kell's essay on types in my thesis; would cite this too if I hadn't just submitted.

**Comment by tpetricek** [comment]
Hello everyone! 

Thanks for the nice words about the essay and for the clarification regarding Julia. I'll make sure to clarify this in the final version (if it gets accepted) or in the next version. Using types _"in all the ways that don't end with the programmer arguing with the compiler"_ is a fantastic summary :-).

I have a feeling that Julia is using types in very interesting ways that completely conflict with many people's assumptions about how types should be used (which makes it a very interesting case!) so I'll definitely try to understand this better - thanks for the references!

**Comment by jiahao** [comment]
+1

I thought the article was nicely written too, but I forgot to put that in the issue. Sorry!
